File: problems/CSES/Graph_Algorithms/monsters/monsters.py
import sys
from collections import deque
# sys.setrecursionlimit(200000)

input = sys.stdin.readline

def main():
    INF = 10 ** 9

    n, m = map(int, input().split())
    grid = [list(input().strip()) for _ in range(n)]

    monster_dist = [[INF] * m for _ in range(n)]
    player_dist = [[-1] * m for _ in range(n)]

    parent = [[None] * m for _ in range(n)]

    q = deque()

    start = None

    for r in range(n):
        for c in range(m):
            if grid[r][c] == 'M':
                monster_dist[r][c] = 0
                q.append((r, c))
            elif grid[r][c] == 'A':
                start = (r, c)

    directions = [
        (1, 0, 'D'),
        (-1, 0, 'U'),
        (0, 1, 'R'),
        (0, -1, 'L')
    ]

    while q:
        r, c = q.popleft()

        for dr, dc, _ in directions:
            nr = r + dr
            nc = c + dc

            if 0 <= nr < n and 0 <= nc < m and grid[nr][nc] != '#' and monster_dist[nr][nc] == INF:
                monster_dist[nr][nc] = monster_dist[r][c] + 1
                q.append((nr, nc))

    sr, sc = start

    if sr == 0 or sr == n - 1 or sc == 0 or sc == m - 1:
        print("YES")
        print(0)
        print()
        sys.exit()

    q = deque([(sr, sc)])
    player_dist[sr][sc] = 0

    end = None

    while q:
        r, c = q.popleft()

        for dr, dc, move in directions:
            nr = r + dr
            nc = c + dc

            if not (0 <= nr < n and 0 <= nc < m):
                continue

            if grid[nr][nc] == '#':
                continue

            if player_dist[nr][nc] != -1:
                continue

            new_dist = player_dist[r][c] + 1

            if new_dist >= monster_dist[nr][nc]:
                continue

            player_dist[nr][nc] = new_dist

            parent[nr][nc] = (r, c, move)

            if nr == 0 or nr == n - 1 or nc == 0 or nc == m - 1:
                end = (nr, nc)
                break

            q.append((nr, nc))

        if end is not None:
            break

    if end is None:
        print("NO")
    else:
        path = []

        r, c = end

        while (r, c) != (sr, sc):
            pr, pc, move = parent[r][c]
            path.append(move)
            r, c = pr, pc

        path.reverse()
        path = ''.join(path)

        print("YES")
        print(len(path))
        print(path)

main()

File: problems/CSES/Graph_Algorithms/monsters/monsters.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

const int INF = 1e9;

const int dr[4] = {1, -1, 0, 0};
const int dc[4] = {0, 0, 1, -1};
const char moves_dir[4] = {'D', 'U', 'R', 'L'};

struct Step {
    int r, c;
    char move;
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    cin >> n >> m;

    vector<string> grid(n);
    vector<vector<int>> monster_dist(n, vector<int>(m, INF));
    vector<vector<int>> player_dist(n, vector<int>(m, -1));
    vector<vector<Step>> parent(n, vector<Step>(m));

    queue<pair<int, int>> q;
    pair<int, int> start = {-1, -1};

    for (int r = 0; r < n; ++r) {
        cin >> grid[r];
        for (int c = 0; c < m; ++c) {
            if (grid[r][c] == 'M') {
                monster_dist[r][c] = 0;
                q.push({r, c});
            } else if (grid[r][c] == 'A') {
                start = {r, c};
            }
        }
    }

    while (!q.empty()) {
        auto [r, c] = q.front();
        q.pop();

        for (int i = 0; i < 4; ++i) {
            int nr = r + dr[i];
            int nc = c + dc[i];

            if (nr >= 0 && nr < n && nc >= 0 && nc < m && 
                grid[nr][nc] != '#' && monster_dist[nr][nc] == INF) {
                monster_dist[nr][nc] = monster_dist[r][c] + 1;
                q.push({nr, nc});
            }
        }
    }

    auto [sr, sc] = start;

    if (sr == 0 || sr == n - 1 || sc == 0 || sc == m - 1) {
        cout << "YES\n0\n\n";
        return 0;
    }

    queue<pair<int, int>> player_q;
    player_q.push({sr, sc});
    player_dist[sr][sc] = 0;

    pair<int, int> end_node = {-1, -1};

    while (!player_q.empty()) {
        auto [r, c] = player_q.front();
        player_q.pop();

        bool reached_exit = false;

        for (int i = 0; i < 4; ++i) {
            int nr = r + dr[i];
            int nc = c + dc[i];

            if (nr < 0 || nr >= n || nc < 0 || nc >= m) continue;
            if (grid[nr][nc] == '#') continue;
            if (player_dist[nr][nc] != -1) continue;

            int new_dist = player_dist[r][c] + 1;
            if (new_dist >= monster_dist[nr][nc]) continue;

            player_dist[nr][nc] = new_dist;
            parent[nr][nc] = {r, c, moves_dir[i]};

            if (nr == 0 || nr == n - 1 || nc == 0 || nc == m - 1) {
                end_node = {nr, nc};
                reached_exit = true;
                break;
            }

            player_q.push({nr, nc});
        }

        if (reached_exit) break;
    }

    if (end_node.first == -1) {
        cout << "NO\n";
    } else {
        string path = "";
        int r = end_node.first;
        int c = end_node.second;

        while (r != sr || c != sc) {
            Step p = parent[r][c];
            path += p.move;
            r = p.r;
            c = p.c;
        }

        reverse(path.begin(), path.end());

        cout << "YES\n" << path.size() << "\n" << path << "\n";
    }

    return 0;
}
